hail/src/main/scala/is/hail/types/encoded/EArray.scala

```diff
@@ -102,20 +102,72 @@ final case class EArray(val elementType: EType, override val required: Boolean =
     val array = cb.newLocal[Long]("array", arrayType.allocate(region, len))
     arrayType.storeLength(cb, array, len)
 
-    val i = cb.newLocal[Int]("i")
     val readElemF = elementType.buildInplaceDecoder(arrayType.elementType, cb.emb.ecb)
+    val pastLastFull64BlockOff = cb.memoize(arrayType.elementOffset(array, len, len - 64))
+    val pastLastOff = cb.memoize(arrayType.elementOffset(array, len, len))
 
-    if (!elementType.required)
+    if (elementType.required) {
+      // val i = cb.newLocal[Int]("i")
+      // cb.forLoop(cb.assign(i, 0), i < len, cb.assign(i, i + 1), {
+      //   val elemOff = cb.memoize(arrayType.elementOffset(array, len, i))
+      //   readElemF(cb, region, elemOff, in)
+      // })
+      //
+      // cb.forLoop({}, elemOff < pastLastFull64BlockOff, {}, {
+      //   for (_ <- 0 to 63) {
+      //     readElemF(cb, region, elemOff, in)
+      //     cb.assign(elemOff, arrayType.nextElementOffset(elemOff))
+      //   }
+      // })
+      val elemOff = cb.newLocal[Long]("elemOff", arrayType.firstElementOffset(array, len))
+      cb.forLoop({}, elemOff < pastLastOff, cb.assign(elemOff, arrayType.nextElementOffset(elemOff)), {
+        readElemF(cb, region, elemOff, in)
+      })
+    } else {
       cb += in.readBytes(region, array + const(arrayType.lengthHeaderBytes), arrayType.nMissingBytes(len))
 
-    cb.forLoop(cb.assign(i, 0), i < len, cb.assign(i, i + 1), {
-      val elemAddr = cb.memoize(arrayType.elementOffset(array, len, i))
-      if (elementType.required)
-        readElemF(cb, region, elemAddr, in)
-      else
+      def unsetRightMostBit(x: Value[Long]): Code[Long] =
+        x & (x - 1)
+
+      val presentBits = cb.newLocal[Long]("presentBits", 0L)
+      val mbyteOffset = cb.newLocal[Long]("mbyteOffset", array + arrayType.lengthHeaderBytes)
+      val blockOff = cb.newLocal[Long]("blockOff", arrayType.firstElementOffset(array, len))
+      val inBlockIndexToPresentValue = cb.newLocal[Int]("inBlockIndexToPresentValue", 0)
+
+      cb.forLoop(
+        {},
+        blockOff < pastLastFull64BlockOff,
+        cb.assign(blockOff, arrayType.incrementElementOffset(blockOff, 64)),
+        {
+          cb.assign(presentBits, ~Region.loadLong(mbyteOffset).reverseBytes)
+          cb.assign(mbyteOffset, mbyteOffset + 8)
+          cb.whileLoop(presentBits.cne(0L), {
+            cb.assign(inBlockIndexToPresentValue, presentBits.numberOfTrailingZeros)
+            val elemOff = cb.memoize(
+              arrayType.incrementElementOffset(blockOff, inBlockIndexToPresentValue))
+            readElemF(cb, region, elemOff, in)
+            cb.assign(presentBits, unsetRightMostBit(presentBits))
+          })
+        }
+      )
+      val i = cb.newLocal[Int]("i", len - (len % 64))
```

Comment:
It should also be possible to avoid the separate loop for the last block. Here's an untested attempt. The idea is to compute a mask which usually leaves all bits unchanged, but on the last incomplete block, zeroes bits which are past the end. Since the inner loop only writes to positions with a set bit, that ensures we handle the last block correctly.
```scala
      val partialBlockSize = cb.memoize(len % 64)
      val partialBlockStart = cb.memoize(arrayType.elementOffset(array, len, len - partialBlockSize)
      val partialBlockMask = cb.memoize(-1 << partialBlockSize)
      val mask = cb.newLocal[Long]("mask", 0L)
      val presentBits = cb.newLocal[Long]("presentBits", 0L)
      val mbyteOffset = cb.newLocal[Long]("mbyteOffset", array + arrayType.lengthHeaderBytes)
      val blockOff = cb.newLocal[Long]("blockOff", arrayType.firstElementOffset(array, len))
      val inBlockIndexToPresentValue = cb.newLocal[Int]("inBlockIndexToPresentValue", 0)

      cb.forLoop(
        {},
        blockOff < pastLastOff,
        cb.assign(blockOff, arrayType.incrementElementOffset(blockOff, 64)),
        {
          // hopefully generates a conditional move, could also try blockOff.ceq(partialBlockStart).toL * partialBlockMask
          cb.ifx(blockOff.ceq(partialBlockStart), cb.assign(mask, partialBlockMask, 0L))
          cb.assign(presentBits, ~(Region.loadLong(mbyteOffset).reverseBytes | mask))
          cb.assign(mbyteOffset, mbyteOffset + 8)
          cb.whileLoop(presentBits.cne(0L), {
            cb.assign(inBlockIndexToPresentValue, presentBits.numberOfTrailingZeros)
            val elemOff = cb.memoize(
              arrayType.incrementElementOffset(blockOff, inBlockIndexToPresentValue))
            readElemF(cb, region, elemOff, in)
            cb.assign(presentBits, unsetRightMostBit(presentBits))
          })
        }
      )
```